### PR TITLE
[ Platform] [#5535] Remove stale instance configs after cloud provider is deleted

### PR DESCRIPTION
**Pull request by jaydeepkumara**
After an onprem cloud-config is deleted we still retain the instance type that can't be deleted. It's misleading when a user is trying to create a universe and this stale entry will be presented as an option to choose as an instance type.

Environment details
Version: 2.2.0.0
Cloud provider: On-prem non replicated
Steps to reproduce

Create an on-prem cloud provider with a single node
Add an instance type in the cloud provider config
Delete the cloud provider
Create a new cloud provider
Goto instance types you will notice the instance type from point 2 shows up in the new cloud provider

Changes:- 
Backend:- Added delete instance type by the provider.
Front end:- No Changes Required for this PR. Page was breaking when we add a node in on-premise and then try to edit on-premise.

Testcase:-
[info] Instrumenting 1219 classes to /Users/jaahir/Documents/yugabyte-db/managed/target/scala-2.11/jacoco/instrumented-classes
[info] Test run started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateProviderWithHostVpcGcp started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testDeleteProviderWithUniverses started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testGcpBootstrapMultiRegionSomeRegionInput started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testEditProviderKubernetes started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateProvider started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateKubernetesMultiRegionProviderFailure started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testListEmptyProviders started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testEditProviderWithEmptyHostedZoneId started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testEditProviderWithAWSProviderType started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testDeleteProviderWithInvalidProviderUUID started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testEditProviderWithInvalidProviderType started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testEditProviderKubernetesConfigEdit started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testDeleteProviderWithMultiRegionAccessKey started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateProviderWithConfig started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testListProvidersWithValidCustomer started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testAwsBootstrapWithDestVpcId started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testListProviders started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateKubernetesMultiRegionProvider started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testDeleteProviderWithoutAccessKey started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testDeleteProviderWithProvisionScript started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testDeleteProviderWithAccessKey started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateAwsProviderWithValidHostedZoneId started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateAwsProviderWithInvalidDevopsReply started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateAwsProviderWithInvalidHostedZoneId started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateWithInvalidParams started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testGcpBootstrapMultiRegionNoRegionInput started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateDuplicateProvider started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateProviderWithHostCredentialsGcp started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testCreateProviderWithDifferentCustomer started
[info] Test com.yugabyte.yw.controllers.CloudProviderControllerTest.testDeleteProviderWithIntsnceType started
[info] Test run finished: 0 failed, 0 ignored, 30 total, 25.736s
[info] Passed: Total 30, Failed 0, Errors 0, Passed 30
[success] Total time: 30 s, completed Mar 1, 2021 1:24:30 PM

### CODE REVIEW

File: managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java [hunk 1]
```diff
@@ -134,6 +141,7 @@ public Result delete(UUID customerUUID, UUID providerUUID) {
         accessKey.delete();
       }
       NodeInstance.deleteByProvider(providerUUID);
+      InstanceType.deleteInstanceTypesForProvider(provider, config);
```
Comment (Arnav15) [inline managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java:144]: So, the way we assign instance types to providers is by using the provider code, and not the provider uuid. So while for most clouds, we just allow one provider, this is not true in the case of k8s. For k8s, we can have multiple providers, and they all share instance types. So, if we always just delete the instances, some providers will end up without associated instance types. And eventually, we will be allowing multiple providers for all clouds. So in this change, first verify that there is no other provider with the same code. If there is, do not delete the instances, and if there isn't, then you can go ahead and delete the instances.

**Review by Arnav15 — CHANGES_REQUESTED** [review]

File: managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java (hunk 1 above)
Comment (Arnav15) [reply managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java:144]: Also, update the description of the issue as mentioned. Add sections for
Summary:
Testing:
(and then the output of the **full** sbt test command)

File: managed/ui/src/components/config/OnPrem/OnPremConfiguration.js [hunk 2]
```diff
@@ -200,8 +200,9 @@ export default class OnPremConfiguration extends Component {
           // Update numNodesConfigured until done
           const numNodesConfigured = this.state.numNodesConfigured + Object.keys(response).length;
           this.setState({ numNodesConfigured: numNodesConfigured });
+          let nodeLength = config.nodes ? config.nodes.length : 0;
           // Launch configuration of access keys once all node instances are bootstrapped
-          if (numNodesConfigured === config.nodes.length) {
+          if (numNodesConfigured === nodeLength) {
```
Comment (sshev) [inline managed/ui/src/components/config/OnPrem/OnPremConfiguration.js:205]:
```suggestion
          if (numNodesConfigured === config.nodes?.length) {
```

File: managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java [hunk 3]
```diff
@@ -134,6 +141,14 @@ public Result delete(UUID customerUUID, UUID providerUUID) {
         accessKey.delete();
       }
       NodeInstance.deleteByProvider(providerUUID);
+
+      int providersCount = Provider.getbyCode(provider.code).size();
```
Comment (SergeyPotachev) [inline managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java:145]: getByCode()

File: managed/src/main/java/com/yugabyte/yw/models/Provider.java [hunk 4]
```diff
@@ -193,6 +193,16 @@ public String getAwsHostedZoneName() {
     return getConfig().get("AWS_HOSTED_ZONE_NAME");
   }
 
+
+  /**
+   * Get all Providers by code without customer uuid.
+   * @param code
+   * @return
+   */
+  public static List<Provider> getbyCode(String code) {
```
Comment (SergeyPotachev) [inline managed/src/main/java/com/yugabyte/yw/models/Provider.java:202]: Please rename to getByCode(...)

File: managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java [hunk 5]
```diff
@@ -55,6 +56,12 @@
 import static com.yugabyte.yw.models.helpers.CommonUtils.DEFAULT_YB_HOME_DIR;
 
 public class CloudProviderController extends AuthenticatedController {
+  private final Config config;
+  @Inject
```
Comment (SergeyPotachev) [inline managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java:60]: Insert empty string before @Inject. Each function should be separated from fields and other functions by an empty line.

File: managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java [hunk 6]
```diff
@@ -134,6 +141,14 @@ public Result delete(UUID customerUUID, UUID providerUUID) {
         accessKey.delete();
       }
       NodeInstance.deleteByProvider(providerUUID);
+
+      int providersCount = Provider.getbyCode(provider.code).size();
+      // Delete Intance type only if 1 provider with code.
+      // More than 1 provider can have shared instance type so we can't delete it.
+      // as instance type are shred acrossed providers
+      if (providersCount == 1) {
```
Comment (SergeyPotachev) [inline managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java:149]: I like this code but I don't like the comment - it is written slightly casually (a lot of typos/errors). Please remember that this code is open source so the text could be read by other people.

// Delete Instance type only if 1 provider with such code exists.
// More than 1 provider can have this instance type so we can't delete it
// if it is shared across providers.

(Something like this, but it is not a standard. Check comments carefully, please.)

File: managed/src/test/java/com/yugabyte/yw/controllers/CloudProviderControllerTest.java [hunk 7]
```diff
@@ -415,6 +422,36 @@ public void testDeleteProviderWithAccessKey() {
     assertAuditEntry(1, customer.uuid);
   }
 
+  @Test
+  public void testDeleteProviderWithIntsnceType() {
```
[the remark this answers is not in the document]
Comment (SergeyPotachev) [reply managed/src/test/java/com/yugabyte/yw/controllers/CloudProviderControllerTest.java:426]: Still not fixed - what is 'IntsnceType' ???

[the remark this answers is not in the document]
Comment (jaydeepkumara) [reply managed/src/test/java/com/yugabyte/yw/controllers/CloudProviderControllerTest.java:426]: As per the existing code, we are not deleting IntsnceType.
There was no test case where we have attached any instance type to any provider while deleting so I have added one

[the remark this answers is not in the document]
Comment (SergeyPotachev) [reply managed/src/test/java/com/yugabyte/yw/controllers/CloudProviderControllerTest.java:426]: I'm simply asking to rename from testDeleteProviderWithIntsnceType to testDeleteProviderWith**Instance**Type.

[the remark this answers is not in the document]
Comment (jaydeepkumara) [reply managed/src/test/java/com/yugabyte/yw/controllers/CloudProviderControllerTest.java]: Done

File: managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java (hunk 6 above)
Comment (jaydeepkumara) [reply managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java:149]: Updated comment let me know if looks fine now.

File: managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java (hunk 5 above)
Comment (jaydeepkumara) [reply managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java:61]: done

**Review by SergeyPotachev — APPROVED** [review]
Quite simple idea and changes. Looks good to me.

**Review by Arnav15 — APPROVED** [review]

**Pull request merged**